lib/kvdb/sched_sts.c

```diff
@@ -1,473 +1,84 @@
-
-    while (true) {
-
-        m_lock(self);
-
-        /* Check SS_EXIT before waiting makes SS_EXIT more responsive,
-         * which has the nice benefit of speeding up unit tests.
-         */
-        ss = atomic_read(&self->state);
-        if (ss != SS_EXIT) {
-            rc = cv_timedwait(&self->mcv, &self->mlock, timeout_ms);
-            assert(rc == 0 || rc == ETIMEDOUT);
-            if (rc == 0)
-                dbg_wake_signals++;
-            else if (rc == ETIMEDOUT)
-                dbg_wake_timeouts++;
-            else
-                dbg_wake_other++;
-        }
-        m_unlock(self);
-
-        ss = atomic_read(&self->state);
-
-        if (csched_rp_dbg_mon(self->rp)) {
-            u64 now = get_time_ns();
-            u64 delay = 5ULL * 1000 * 1000 * 1000;
-
-            if (now > dbg_last + delay) {
-                log_info("sts/mon sig %lu timeo %lu other %lu ss %d",
-                         (ulong)dbg_wake_signals,
-                         (ulong)dbg_wake_timeouts,
-                         (ulong)dbg_wake_other,
-                         ss);
-
-                dbg_wake_signals = 0;
-                dbg_wake_timeouts = 0;
-                dbg_wake_other = 0;
-                dbg_last = now;
-            }
-        }
-
-        if (ss == SS_EXIT)
-            break;
-
-        timeout_ms = MONITOR_TIMEOUT_MS;
-
-        if (add_worker_next_ns <= get_time_ns()) {
-            merr_t err;
-            uint   i;
-
-            for (i = 0; i <= self->qc; i++) {
-                if (sts_worker_surplus(self, i) < 0) {
-                    err = add_worker(self, i);
-                    if (!err) {
-                        add_worker_next_ns = 0;
-                        timeout_ms = 10;
-                    } else {
-                        add_worker_next_ns = get_time_ns() + 5 * NSEC_PER_SEC;
-                    }
-                }
-            }
-        }
-    }
-
-    if (csched_rp_dbg_mon(self->rp))
-        log_info("sts/mon exiting, waiting for %d workers to finish",
-                 atomic_read(&self->spawned_workers));
-
-    while (atomic_read(&self->spawned_workers) > 0)
-        msleep(EXIT_SLEEP_MS);
-
-    cancel_jobs_all(self);
-
-    return 0;
-}
-
-static void
-sts_perfc_alloc_internal(struct sts *self)
+sts_job_submit(struct sts *self, struct sts_job *job)
 {
-    char namebuf[32], qnum_str[16];
-    int  i;
-
-    snprintf(namebuf, sizeof(namebuf), "kvdb/%s", self->name);
-
-    /* Iterate w/ <= to get shared worker stats */
-    for (i = 0; i <= self->qc; i++) {
-        snprintf(qnum_str, sizeof(qnum_str), "q%d", i);
-        sts_perfc_alloc(self->rp->perfc_level, namebuf, qnum_str, &self->qv[i].qpc);
-    }
-}
+    assert(self && job);
+    assert(job->sj_job_fn);
 
-static void
-sts_perfc_free_internal(struct sts *self)
-{
-    int i;
+    if (csched_rp_dbg_jobs(self->sts_rp))
+        log_info("sts/job %u submit qnum %u", job->sj_id, job->sj_qnum);
```

Comment:
Maybe get rid of this and other debug statements (there are only 2).  Then drop all knowledge of `kvdb_rparams` from this file.

Reply:
will do